tests/ui/unnecessary_literal_unwrap.rs

```diff
@@ -0,0 +1,49 @@
+// run-rustfix
+#![warn(clippy::unnecessary_literal_unwrap)]
+#![allow(unreachable_code)]
+#![allow(clippy::unnecessary_lazy_evaluations)]
+
+fn unwrap_option_some() {
+    let _val = Some(1).unwrap();
+    let _val = Some(1).expect("this never happens");
+}
+
+fn unwrap_option_none() {
+    None::<usize>.unwrap();
+    None::<usize>.expect("this always happens");
+}
+
+fn unwrap_result_ok() {
+    let _val = Ok::<usize, ()>(1).unwrap();
```

Comment:
What if the `usize` was needed in this context? Wouldn't suggesting `let _val = 1;` loose that information? Could you include examples in which the type information is needed, such as:

```rust
fn context_needed() {
    let _s = Some::<usize>([1, 2, 3].iter().sum()).unwrap();
}
```

as the current suggestion will not compile as-is.

Reply:
Agreed. If the `Some` or `Ok` have given generics (from the hir::Ty, not the ty::Ty), the first of them should be added to the suggestion as type of the binding.

Reply:
I am stuck trying to address the review comments. I have an `Expr` from `Expr::call` [here](https://github.com/rust-lang/rust-clippy/pull/10358/files#diff-bf1cd10676ad7ed2fa17ccd518ddaaba2f1a54859f8cbd9f6745d87bc9ef99adR18), but how do I get the generic arguments (`hir::Ty`) from that?